source/extensions/filters/http/rate_limit_quota/quota_bucket_cache.h

```diff
@@ -0,0 +1,119 @@
+#pragma once
+#include <algorithm>
+#include <memory>
+
+#include "envoy/extensions/filters/http/rate_limit_quota/v3/rate_limit_quota.pb.h"
+#include "envoy/extensions/filters/http/rate_limit_quota/v3/rate_limit_quota.pb.validate.h"
+#include "envoy/service/rate_limit_quota/v3/rlqs.pb.h"
+#include "envoy/service/rate_limit_quota/v3/rlqs.pb.validate.h"
+
+#include "source/common/protobuf/utility.h"
+#include "source/extensions/filters/http/common/factory_base.h"
+#include "source/extensions/filters/http/rate_limit_quota/client.h"
+
+#include "absl/time/time.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace RateLimitQuota {
+
+using ::envoy::service::rate_limit_quota::v3::BucketId;
+using ::envoy::service::rate_limit_quota::v3::RateLimitQuotaUsageReports;
+
+using BucketAction = ::envoy::service::rate_limit_quota::v3::RateLimitQuotaResponse::BucketAction;
+using BucketQuotaUsage =
+    ::envoy::service::rate_limit_quota::v3::RateLimitQuotaUsageReports::BucketQuotaUsage;
+
+// Customized hash and equal struct for `BucketId` hash key.
+struct BucketIdHash {
+  size_t operator()(const BucketId& bucket_id) const { return MessageUtil::hash(bucket_id); }
```

Comment:
This is where the downside of using protos as data types will come into play. The `MessageUtil::hash` is very expensive, as it converts proto into JSON string and then computes its hash. We had to reduce its use in xDS updates, since they consumed noticeable amount of CPU.

Same is for the BucketIdEqual below. The MessageDifferencer is very expensive as it compares fields one by one.